nowcasting_dataset/manager.py

```diff
@@ -0,0 +1,384 @@
+                locations_for_each_example["t0_datetime_UTC"]
+            )
+            locations_for_each_example_of_each_split[split_name] = locations_for_each_example
+
+        # Fire up a separate process for each DataSource, and pass it a list of batches to
+        # create, and whether to utils.upload_and_delete_local_files().
+        # TODO: Split this up into separate functions!!!
+        n_data_sources = len(self.data_sources)
+        nd_utils.set_fsspec_for_multiprocess()
+        for split_name in splits_which_need_more_batches:
+            locations_for_split = locations_for_each_example_of_each_split[split_name]
+            with futures.ProcessPoolExecutor(max_workers=n_data_sources) as executor:
+                future_create_batches_jobs = []
+                for worker_id, (data_source_name, data_source) in enumerate(
+                    self.data_sources.items()
+                ):
+                    # Get indexes of first batch and example. And subset locations_for_split.
+                    idx_of_first_batch = first_batches_to_create[split_name][data_source_name]
+                    idx_of_first_example = idx_of_first_batch * self.config.process.batch_size
+                    locations = locations_for_split.loc[idx_of_first_example:]
+
+                    # Get paths.
+                    dst_path = (
+                        self.config.output_data.filepath / split_name.value / data_source_name
+                    )
+                    local_temp_path = (
+                        self.local_temp_path
+                        / split_name.value
+                        / data_source_name
+                        / f"worker_{worker_id}"
+                    )
+
+                    # Make folders.
+                    nd_fs_utils.makedirs(dst_path, exist_ok=True)
+                    if self.save_batches_locally_and_upload:
+                        nd_fs_utils.makedirs(local_temp_path, exist_ok=True)
+
+                    # Submit data_source.create_batches task to the worker process.
+                    future = executor.submit(
+                        data_source.create_batches,
+                        spatial_and_temporal_locations_of_each_example=locations,
+                        idx_of_first_batch=idx_of_first_batch,
+                        batch_size=self.config.process.batch_size,
+                        dst_path=dst_path,
+                        local_temp_path=local_temp_path,
+                        upload_every_n_batches=self.config.process.upload_every_n_batches,
+                    )
+                    future_create_batches_jobs.append(future)
+
+                # Wait for all futures to finish:
+                for future in future_create_batches_jobs:
+                    # Call exception() to propagate any exceptions raised by the worker process into
+                    # the main process, and to wait for the worker to finish.
+                    future.exception()
```

Comment:
if this takes a long time, is it worth having some sort of logging saying how far through we are?

Reply:
Good point!  I've added some logging to `DataSource.create_batches()`.